controllers/nodehealthcheck_controller.go

```diff
@@ -353,60 +350,116 @@ func (r *NodeHealthCheckReconciler) remediate(node *v1.Node, nhc *remediationv1a
+			return nil, nil
+		}
+		return nil, errors.Wrapf(err, "failed to get current template")
+	}
+	remediationCR, err := rm.GenerateRemediationCR(node, nhc, currentTemplate)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to generate remediation CR")
+	}
+
 	if isControlPlaneNode {
 		labels := remediationCR.GetLabels()
 		labels[RemediationControlPlaneLabelKey] = ""
 		remediationCR.SetLabels(labels)
 	}
 
-	if created, err := rm.CreateRemediationCR(remediationCR, nhc); err != nil {
+	// create remediation CR
+	created, err := rm.CreateRemediationCR(remediationCR, nhc)
+	if err != nil {
 		return nil, errors.Wrapf(err, "failed to create remediation CR")
-	} else if created {
+	}
+
+	// always update status, in case patching it failed during last reconcile
+	resources.UpdateStatusRemediationStarted(node, nhc, remediationCR)
+
+	if created {
 		r.Recorder.Event(nhc, eventTypeNormal, eventReasonRemediationCreated, fmt.Sprintf("Created remediation object for node %s", node.Name))
+		var requeueIn *time.Duration
+		if timeout != nil {
+			// come back when timeout expires
+			requeueIn = pointer.Duration(timeout.Duration + 1*time.Second)
+		}
+		return requeueIn, nil
+	}
+
+	// CR already exists, check for timeout in case we need to
+	if timeout == nil {
+		// no timeout set for classic remediation
+		// nothing to do anymore here
+		return nil, nil
+	}
+
+	startedRemediation := resources.FindStatusRemediation(node, nhc, func(r *remediationv1alpha1.Remediation) bool {
+		return r.Resource.GroupVersionKind() == remediationCR.GroupVersionKind()
+	})
+
+	if startedRemediation == nil {
+		// should not have happened, seems last status update failed
+		// retry asap
+		return pointer.Duration(1 * time.Second), nil
+	}
+
+	if startedRemediation.TimedOut != nil {
+		// timeout handled already: should not have happened, but ok. Just reconcile again asap for trying the next template
+		return pointer.Duration(1 * time.Second), nil
 	}
 
-	isAlert, nextReconcile := r.alertOldRemediationCR(remediationCR)
-	if isAlert {
-		metrics.ObserveNodeHealthCheckOldRemediationCR(node.Name, node.Namespace)
+	now := metav1.Now()
+	timeoutAt := startedRemediation.Started.Add(timeout.Duration)
+	if !now.After(timeoutAt) {
+		// not timed out yet, come back when we do so
+		return pointer.Duration(timeoutAt.Sub(now.Time)), nil
 	}
-	return nextReconcile, nil
+
+	// handle timeout
+
+	// add timeout annotation to remediation CR
+	annotations := remediationCR.GetAnnotations()
+	if annotations == nil {
+		annotations = make(map[string]string, 1)
+	}
+	annotations[remediationTimedOutAnnotationkey] = metav1.Now().Format(time.RFC3339)
+	remediationCR.SetAnnotations(annotations)
+	if rm.UpdateRemediationCR(remediationCR); err != nil {
+		return nil, errors.Wrapf(err, "failed to update remediation CR with timeout annotation")
+	}
+
+	// update status (important to do this after CR update, else we won't retry that update in case of error)
+	startedRemediation.TimedOut = &now
```

Comment:
Can you elaborate on that ? (I'm not completely clear on where do we persist this change)

Reply:
see deferred `patchStatus` call nearly on top of `Reconcile`